vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs

```diff
@@ -1739,49 +1726,66 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
                     VSConstants.S_OK
                 member x.UpdateSolution_Cancel() =
                     VSConstants.S_OK
-                member x.OnActiveProjectCfgChange(pIVsHierarchy) =
-                    OnActiveProjectCfgChange(pIVsHierarchy)
+                member x.OnActiveProjectCfgChange(pHierProj) =
+                    UpdateConfig(pHierProj)
+                    VSConstants.S_OK
                 member x.UpdateProjectCfg_Begin(pHierProj, _pCfgProj, _pCfgSln, _dwAction, pfCancel) =
                     UpdateConfig(pHierProj)
                     VSConstants.S_OK
                 member x.UpdateProjectCfg_Done(pHierProj, _pCfgProj, _pCfgSln, _dwAction, _fSuccess, _fCancel) =
                     UpdateConfig(pHierProj)
+                    // We wait until allproject configurations have been updated before updating our MSBuild state
+                    projNode.UpdateMSBuildState()
+                    projNode.SetProjectFileDirty(projNode.IsProjectFileDirty)
+                    projNode.ComputeSourcesAndFlags()
```

Comment:
Here's what I see for the set of events triggered when loading ``tests\projects\misc\TestProjectChanges.sln``. Looking at this, 


* I believe you are right that we are doing ComputeSourcesAndFlags twice per project but only because we are getting two complete sets of ``OnActiveProjectCfgChangeBatchBegin``/ ``OnActiveProjectCfgChangeBatchEnd`` events sent through.

* It is important not to do any ComputeSourcesAndFlags until we have updated the configuration on all the projects. So I think we have to do this in OnActiveProjectCfgChangeBatchEnd.

* The cost of ``ComputeSourcesAndFlags`` has been much reduced since #3002 since it uses the stored computed options for referenced projects


### Loading tests\projects\misc\TestProjectChanges.sln

* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3

Then we get a "duplicate" set of Batch events
* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3

### Switching to "Release"
* OnBeforeActiveSolutionCfgChange x 3
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3
* OnAfterActiveSolutionCfgChange x 3

### On prompted solution reload after a project file has been edited
* OnActiveProjectCfgChangeBatchBegin x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(null) x 3
* IVsUpdateSolutionEvents.OnActiveProjectCfgChange(not-null) x 6
* OnActiveProjectCfgChangeBatchEnd x 3

Then we get a "duplicate" set of Batch events

* OnActiveProjectCfgChangeBatchBegin x 3
* OnActiveProjectCfgChangeBatchEnd x 3

### On individual project reload 

* OnActiveProjectCfgChange 

Also I never see these being called:
* IVsUpdateSolutionEvents2.OnActiveProjectCfgChange
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Begin
* IVsUpdateSolutionEvents2.UpdateProjectCfg_Done